### PR TITLE
Add HasMapping to sql adapter

### PR DESCRIPTION
**Pull request by oatovar**
* Add ability to check if an alias exists within the user's short link repo.
* Add integration tests to cover FindAliasByUser

This is the first PR related to #798. It updates the sql adapter so that it can verify if a specific alias is owned by a given user.

### CODE REVIEW

File: backend/app/adapter/sqldb/user_short_link.go [hunk 1]
```diff
@@ -63,6 +63,24 @@ func (u UserShortLinkSQL) FindAliasesByUser(user entity.User) ([]string, error)
 	return aliases, nil
 }
 
+// IsAliasOwnedByUser verifies if the ShortLink alias belongs to the given user.
+func (u UserShortLinkSQL) IsAliasOwnedByUser(user entity.User, alias string) (bool, error) {
+	statement := fmt.Sprintf(`SELECT from "%s" WHERE "%s"=$1 AND "%s"=$2`,
```
Comment (arberiii) [inline backend/app/adapter/sqldb/user_short_link.go:68]:
```suggestion
	statement := fmt.Sprintf(`SELECT FROM "%s" WHERE "%s"=$1 AND "%s"=$2`,
```

File: backend/app/adapter/sqldb/user_short_link_integration_test.go [hunk 2]
```diff
@@ -114,6 +114,128 @@ func TestListShortLinkSql_FindAliasesByUser(t *testing.T) {
 	}
 }
 
+func TestListShortLinkSql_IsAliasOwnedByUser(t *testing.T) {
+	now := mustParseTime(t, "2019-05-01T08:02:16Z")
+	user := entity.User{
```
Comment (arberiii) [inline backend/app/adapter/sqldb/user_short_link_integration_test.go:119]: You can directly put the user into the test struct. Also fields LastSignedInAt, CreatedAt, UpdatedAt are redundant.

File: backend/app/adapter/sqldb/user_short_link_integration_test.go [hunk 3]
```diff
@@ -114,6 +114,128 @@ func TestListShortLinkSql_FindAliasesByUser(t *testing.T) {
...
+				dbMigrationRoot,
+				dbConfig,
+				func(sqlDB *sql.DB) {
+					insertUserTableRows(t, sqlDB, testCase.userTableRows)
+					insertShortLinkTableRows(t, sqlDB, testCase.shortLinkTableRows)
+					insertUserShortLinkTableRows(t, sqlDB, testCase.relationTableRows)
+
+					userShortLinkRepo := sqldb.NewUserShortLinkSQL(sqlDB)
+					result, err := userShortLinkRepo.IsAliasOwnedByUser(testCase.user, testCase.alias)
+					assert.Equal(t, nil, err)
```
Comment (arberiii) [inline backend/app/adapter/sqldb/user_short_link_integration_test.go:232]: Out of context but how do you actually run these tests? I tried with scripts/integration-test but even if I put an assert.Equal(t, 1, 2) test passes

File: backend/app/adapter/sqldb/user_short_link_integration_test.go (hunk 2 above)
Comment (oatovar) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:119]: I'm a bit confused as to the reasoning behind this. Writing the user in the test struct multiple times would just create duplicated code.

Comment (oatovar) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:119]: As far as the time stamps go, they're set so that there is no nil pointer left which can lead to errors down the road 😄

Comment (arberiii) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:119]: I was suggesting for easy readability but we can leave it as it is since it is readable also this way. Good point about timestamps, if we change functioning, those tests should not fail for nil pointers!

Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:119]: @oatovar Making the test case easy to follow avoid distracting engineers reading the code away from coming up edge case.

File: backend/app/adapter/sqldb/user_short_link.go [hunk 4]
```diff
@@ -63,6 +63,24 @@ func (u UserShortLinkSQL) FindAliasesByUser(user entity.User) ([]string, error)
 	return aliases, nil
 }
 
+// IsAliasOwnedByUser verifies if the ShortLink alias belongs to the given user.
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link.go:66]:
```suggestion
// IsAliasOwnedByUser checks whether a given short link belongs to a user.
```

File: backend/app/adapter/sqldb/user_short_link.go [hunk 5]
```diff
@@ -63,6 +63,24 @@ func (u UserShortLinkSQL) FindAliasesByUser(user entity.User) ([]string, error)
...
+func (u UserShortLinkSQL) IsAliasOwnedByUser(user entity.User, alias string) (bool, error) {
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link.go:67]:
```suggestion
func (u UserShortLinkSQL) HasMapping(user entity.User, alias string) (bool, error) {
```

Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link.go:67]: I prefer delegating the concept of ownership to use case.

File: backend/app/adapter/sqldb/user_short_link_integration_test.go (hunk 2 above)
Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:119]: Let's remove the used fields and move user logic into the test case. We may have to modify the test case as the system evolve. Having user in the test case themselves allowing us to make changes without touching the rest.

File: backend/app/adapter/sqldb/user_short_link.go [hunk 6]
```diff
@@ -63,6 +63,24 @@ func (u UserShortLinkSQL) FindAliasesByUser(user entity.User) ([]string, error)
...
+	)
+
+	rows, err := u.db.Query(statement, user.ID, alias)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close()
+
+	found := rows.Next()
+	return found, nil
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link.go:81]: Let's try something like this so that it's more concise and readable:
```go
err := u.db.QueryRow(query, id).Scan(&id)
	if err == sql.ErrNoRows {
		return false, nil
	}
	if err != nil {
		return false, err
	}
```

File: backend/app/adapter/sqldb/user_short_link_integration_test.go [hunk 7]
```diff
@@ -114,6 +114,128 @@ func TestListShortLinkSql_FindAliasesByUser(t *testing.T) {
...
+		name               string
+		userTableRows      []userTableRow
+		shortLinkTableRows []shortLinkTableRow
+		relationTableRows  []userShortLinkTableRow
+		alias              string
+		user               entity.User
+		isFound            bool
+	}{
+		{
+			name: "alias does not exist",
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link_integration_test.go:138]: Let's add a case when user does not exist.

File: backend/app/adapter/sqldb/user_short_link_integration_test.go [hunk 8]
```diff
@@ -114,6 +114,128 @@ func TestListShortLinkSql_FindAliasesByUser(t *testing.T) {
...
+	}
+
+	testCases := []struct {
+		name               string
+		userTableRows      []userTableRow
+		shortLinkTableRows []shortLinkTableRow
+		relationTableRows  []userShortLinkTableRow
+		alias              string
+		user               entity.User
+		isFound            bool
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link_integration_test.go:135]:
```suggestion
		expectIsFound            bool
```

Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:135]: Let's add expect for the expected result so that it's intuitive for readers to figure out the purpose of the test

File: backend/app/adapter/sqldb/user_short_link.go (hunk 5 above)
Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link.go:67]: We may have different algorithm to determine `Ownership` in the future. This leaves us the flexibility to make changes.

File: backend/app/adapter/sqldb/user_short_link_integration_test.go (hunk 7 above)
Comment (oatovar) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:138]: I'm not sure what you mean by this. The relation table cannot have an entry without an exisiting user because of the foreign key usage. Do you mean when the user entity is not set in the test case?

Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link_integration_test.go:130]: @oatovar Awesome! Please ignore this comment then.

File: backend/app/adapter/sqldb/user_short_link.go [hunk 9]
```diff
@@ -63,6 +63,26 @@ func (u UserShortLinkSQL) FindAliasesByUser(user entity.User) ([]string, error)
 	return aliases, nil
 }
 
+// HasMapping checks whether a given short link belongs to a user.
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link.go:66]:
```suggestion
// HasMapping checks whether a given short link is tied to a user.
```

File: backend/app/adapter/sqldb/user_short_link.go [hunk 10]
```diff
@@ -63,6 +63,26 @@ func (u UserShortLinkSQL) FindAliasesByUser(user entity.User) ([]string, error)
...
+
+	var id string
+	err := u.db.QueryRow(query, user.ID, alias).Scan(&id)
+	if err == sql.ErrNoRows {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
```
Comment (byliuyang) [inline backend/app/adapter/sqldb/user_short_link.go:83]: I wonder whether we can do the following:
```go
return true, err
```
When there is err, the return value is ignored.

Comment (oatovar) [reply backend/app/adapter/sqldb/user_short_link.go:83]: Should the value be trusted if there was an error?

Comment (oatovar) [reply backend/app/adapter/sqldb/user_short_link.go:83]: I see what you mean now. I think it would still be better to have it return as is in case someone accidentally uses the returned value.

Comment (byliuyang) [reply backend/app/adapter/sqldb/user_short_link.go:83]: @oatovar It's Go's convention not to use the returned value if there is error. I am fine keeping the code as you have.
